File: src/ast.rs
use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Variable(pub String);

impl From<&str> for Variable {
    fn from(value: &str) -> Self {
        Variable(value.to_string())
    }
}

impl From<String> for Variable {
    fn from(value: String) -> Self {
        Variable(value)
    }
}

impl Display for Variable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Num,
    Bool,
    Unit,
    Var(Variable),
    Fn { arg: Box<Type>, ret: Box<Type> },
    Product { left: Box<Type>, right: Box<Type> },
    Sum { left: Box<Type>, right: Box<Type> },
    Rec { a: Variable, tau: Box<Type> },
    Forall { a: Variable, tau: Box<Type> },
    Exists { a: Variable, tau: Box<Type> },
}

impl std::fmt::Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Num => write!(f, "num"),
            Type::Bool => write!(f, "bool"),
            Type::Unit => write!(f, "()"),
            Type::Var(v) => write!(f, "{}", v),
            Type::Fn { arg, ret } => write!(f, "{} → {}", arg, ret),
            Type::Product { left, right } => write!(f, "{} * {}", left, right),
            Type::Sum { left, right } => write!(f, "{} + {}", left, right),
            Type::Rec { a, tau } => write!(f, "μ {} . {}", a, tau),
            Type::Forall { a, tau } => write!(f, "∀ {} . {}", a, tau),
            Type::Exists { a, tau } => write!(f, "∃ {} . {}", a, tau),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AddOp {
    Add,
    Sub,
}

impl Display for AddOp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AddOp::Add => write!(f, "+"),
            AddOp::Sub => write!(f, "-"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MulOp {
    Mul,
    Div,
}

impl Display for MulOp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MulOp::Mul => write!(f, "*"),
            MulOp::Div => write!(f, "/"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RelOp {
    Lt,
    Gt,
    Eq,
}

impl Display for RelOp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RelOp::Lt => write!(f, "<"),
            RelOp::Gt => write!(f, ">"),
            RelOp::Eq => write!(f, "="),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Direction {
    Left,
    Right,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr {
    Num(i32),
    Addop {
        binop: AddOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Mulop {
        binop: MulOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    True,
    False,
    If {
        cond: Box<Expr>,
        then_: Box<Expr>,
        else_: Box<Expr>,
    },
    Relop {
        relop: RelOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    And {
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Or {
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Var(Variable),
    DeBruijn(usize),
    Lam {
        x: Variable,
        e: Box<Expr>,
    },
    App {
        lam: Box<Expr>,
        arg: Box<Expr>,
    },
    Unit,
    Pair {
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Project {
        e: Box<Expr>,
        d: Direction,
    },
    Inject {
        e: Box<Expr>,
        d: Direction,
    },
    Case {
        e: Box<Expr>,
        xleft: Variable,
        eleft: Box<Expr>,
        xright: Variable,
        eright: Box<Expr>,
    },
    Fix {
        x: Variable,
        e: Box<Expr>,
    },
    TyLam {
        a: Variable,
        e: Box<Expr>,
    },
    TyApp {
        e: Box<Expr>,
        tau: Box<Type>,
    },
    Fold {
        e: Box<Expr>,
        tau: Box<Type>,
    },
    Unfold(Box<Expr>),
    Export {
        e: Box<Expr>,
        tau_adt: Box<Type>,
        tau_mod: Box<Type>,
    },
    Import {
        x: Variable,
        a: Variable,
        e_mod: Box<Expr>,
        e_body: Box<Expr>,
    },
}

impl Display for Expr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Expr::Var(v) => write!(f, "{}", v),
            Expr::DeBruijn(d) => write!(f, "<{}>", d),
            Expr::Num(n) => write!(f, "{}", n),
            Expr::True => write!(f, "true"),
            Expr::False => write!(f, "false"),
            Expr::Unit => write!(f, "()"),
            Expr::Addop { binop, left, right } => write!(f, "({} {} {})", left, binop, right),
            Expr::Mulop { binop, left, right } => write!(f, "({} {} {})", left, binop, right),
            Expr::If { cond, then_, else_ } => {
                write!(f, "(if {} then {} else {})", cond, then_, else_)
            }
            Expr::Relop { relop, left, right } => write!(f, "({} {} {})", left, relop, right),
            Expr::And { left, right } => write!(f, "({} && {})", left, right),
            Expr::Or { left, right } => write!(f, "({} || {})", left, right),
            Expr::Pair { left, right } => write!(f, "({} , {})", left, right),
            Expr::Project { e, d } => match (e.as_ref(), d) {
                (Expr::Pair { left, .. }, Direction::Left) => write!(f, "{}", left),
                (Expr::Pair { right, .. }, Direction::Right) => write!(f, "{}", right),
                _ => write!(f, "{:?}", self),
            },
            Expr::Inject { e, .. } => write!(f, "{}", e),
            Expr::Case {
                e,
                xleft,
                eleft,
                xright,
                eright,
            } => write!(
                f,
                "(case {} of L({}) -> {} | R({}) -> {})",
                e, xleft, eleft, xright, eright
            ),
            Expr::App { lam, arg } => write!(f, "({} {})", lam, arg),
            Expr::Lam { x, e } => write!(f, "(λ {} -> {})", x, e),
            Expr::TyLam { a, e } => write!(f, "(Λ {} -> {})", a, e),
            Expr::TyApp { e, tau } => write!(f, "({} {})", e, tau),
            Expr::Fix { x, e } => write!(f, "(fix {} -> {})", x, e),
            Expr::Fold { e, .. } => write!(f, "(fold {} as ...)", e),
            Expr::Unfold(e) => write!(f, "(unfold {})", e),
            Expr::Export {
                e,
                tau_adt,
                tau_mod,
            } => write!(f, "(export {} without {} as {})", e, tau_adt, tau_mod),
            Expr::Import {
                x,
                a,
                e_mod,
                e_body,
            } => write!(f, "(import ({}, {}) = {} in {})", x, a, e_mod, e_body),
        }
    }
}

File: src/AST.hs
module AST where

data Variable
  = VName String
  | VDeBruijn Int
  deriving (Eq, Ord)

instance Show Variable where
  show (VName n) = n
  show (VDeBruijn n) = "<" ++ show n ++ ">"

data Type
  = TNum
  | TBool
  | TUnit
  | TVar Variable
  | TFn Type Type
  | TProduct Type Type
  | TSum Type Type
  | TForall Variable Type
  | TMu Variable Type
  deriving (Eq)

instance Show Type where
  show TNum = "num"
  show TBool = "bool"
  show TUnit = "()"
  show (TVar v) = show v
  show (TFn arg ret) = "(" ++ show arg ++ " → " ++ show ret ++ ")"
  show (TProduct left right) = show left ++ " * " ++ show right
  show (TSum left right) = show left ++ " + " ++ show right
  show (TForall a tau) = "∀ " ++ show a ++ " . " ++ show tau
  show (TMu a body) = "(μ " ++ show a ++ " . " ++ show body ++ ")"

data AddOp = Add | Sub
  deriving (Eq)

instance Show AddOp where
  show Add = "+"
  show Sub = "-"

data MulOp = Mul | Div
  deriving (Eq)

instance Show MulOp where
  show Mul = "*"
  show Div = "/"

data RelOp = Lt | Gt | Eq
  deriving (Eq)

instance Show RelOp where
  show Lt = "<"
  show Gt = ">"
  show Eq = "="

data Direction = L | R
  deriving (Eq, Show)

data Expr
  = ENum Int
  | EAddop AddOp Expr Expr
  | EMulop MulOp Expr Expr
  | ETrue
  | EFalse
  | EIf Expr Expr Expr
  | ERelop RelOp Expr Expr
  | EAnd Expr Expr
  | EOr Expr Expr
  | EVar Variable
  | ELam Variable (Maybe Type) Expr
  | EApp Expr Expr
  | EUnit
  | EPair Expr Expr
  | EProject Expr Direction
  | EInject Expr Direction (Maybe Type)
  | ECase
      { e :: Expr,
        xleft :: Variable,
        eleft :: Expr,
        xright :: Variable,
        eright :: Expr
      }
  | EFix Variable Expr
  | ELet Variable (Maybe Type) Expr Expr
  deriving (Eq)

instance Show Expr where
  show (EVar v) = show v
  show (ENum n) = show n
  show ETrue = "true"
  show EFalse = "false"
  show EUnit = "()"
  show (EAddop op left right) = "(" ++ show left ++ " " ++ show op ++ " " ++ show right ++ ")"
  show (EMulop op left right) = "(" ++ show left ++ " " ++ show op ++ " " ++ show right ++ ")"
  show (EIf cond then_ else_) = "(if " ++ show cond ++ " then " ++ show then_ ++ " else " ++ show else_ ++ ")"
  show (ERelop op left right) = "(" ++ show left ++ " " ++ show op ++ " " ++ show right ++ ")"
  show (EAnd left right) = "(" ++ show left ++ " && " ++ show right ++ ")"
  show (EOr left right) = "(" ++ show left ++ " || " ++ show right ++ ")"
  show (EPair left right) = "(" ++ show left ++ " , " ++ show right ++ ")"
  show (EProject e d) = case (e, d) of
    (EPair left _, L) -> show left
    (EPair _ right, R) -> show right
    _ -> show e ++ "." ++ show d
  show (EInject e d Nothing) = show e
  show (EInject e d (Just t)) = "(inj " ++ show e ++ " = " ++ show d ++ " as " ++ show t ++ ")"
  show (ECase e xleft eleft xright eright) =
    "(case " ++ show e ++ " of L(" ++ show xleft ++ ") -> " ++ show eleft ++ " | R(" ++ show xright ++ ") -> " ++ show eright ++ ")"
  show (EApp lam arg) = "(" ++ show lam ++ " " ++ show arg ++ ")"
  show (ELam x Nothing e) = "(λ " ++ show x ++ " -> " ++ show e ++ ")"
  show (ELam x (Just t) e) = "(λ " ++ show x ++ " : " ++ show t ++ " -> " ++ show e ++ ")"
  show (ELet x Nothing e_x e_in) = "(let " ++ show x ++ " = " ++ show e_x ++ " in " ++ show e_in ++ ")"
  show (ELet x (Just t) e_x e_in) = "(let " ++ show x ++ " : " ++ show t ++ " = " ++ show e_x ++ " in " ++ show e_in ++ ")"
  show (EFix x e) = "(fix " ++ show x ++ " -> " ++ show e ++ ")"

showSimplified :: Expr -> String
showSimplified (EVar v) = show v
showSimplified (ENum n) = show n
showSimplified ETrue = "true"
showSimplified EFalse = "false"
showSimplified EUnit = "()"
showSimplified (EAddop op left right) = "(" ++ showSimplified left ++ " " ++ show op ++ " " ++ showSimplified right ++ ")"
showSimplified (EMulop op left right) = "(" ++ showSimplified left ++ " " ++ show op ++ " " ++ showSimplified right ++ ")"
showSimplified (EIf cond then_ else_) = "(if " ++ showSimplified cond ++ " then " ++ showSimplified then_ ++ " else " ++ showSimplified else_ ++ ")"
showSimplified (ERelop op left right) = "(" ++ showSimplified left ++ " " ++ show op ++ " " ++ showSimplified right ++ ")"
showSimplified (EAnd left right) = "(" ++ showSimplified left ++ " && " ++ showSimplified right ++ ")"
showSimplified (EOr left right) = "(" ++ showSimplified left ++ " || " ++ showSimplified right ++ ")"
showSimplified (EPair left right) = "(" ++ showSimplified left ++ " , " ++ showSimplified right ++ ")"
showSimplified (EProject e d) = case (e, d) of
  (EPair left _, L) -> showSimplified left
  (EPair _ right, R) -> showSimplified right
  _ -> showSimplified e ++ "." ++ show d
showSimplified (EInject e _ _) = showSimplified e
showSimplified (ECase e xleft eleft xright eright) =
  "(case " ++ showSimplified e ++ " of L(" ++ show xleft ++ ") -> " ++ showSimplified eleft ++ " | R(" ++ show xright ++ ") -> " ++ showSimplified eright ++ ")"
showSimplified (EApp lam arg) = "(" ++ showSimplified lam ++ " " ++ showSimplified arg ++ ")"
showSimplified (ELam x _ e) = "(λ " ++ show x ++ " -> " ++ showSimplified e ++ ")"
showSimplified (ELet x _ e_x e_in) = "(let " ++ show x ++ " = " ++ showSimplified e_x ++ " in " ++ showSimplified e_in ++ ")"
showSimplified (EFix x e) = "(fix " ++ show x ++ " -> " ++ showSimplified e ++ ")"